safehttp/form_test.go

```diff
@@ -0,0 +1,1056 @@
+	for _, test := range tests {
+		for _, req := range *test.reqs {
+			m := NewMachinery(func(rw ResponseWriter, ir *IncomingRequest) Result {
+				form, err := getParsedForm(ir)
+				if err != nil {
+					t.Fatalf(`getParsedForm: got "%v", want nil`, err)
+				}
+				switch got := test.got.(type) {
+				case []int8:
+					form.Slice(&got, "pizza")
+					if diff := cmp.Diff(test.want, got); diff != "" {
+						t.Errorf("form.Slice: got %v, want %v", got, test.want)
+					}
+					if err = form.Err(); test.err.Error() != err.Error() {
+						t.Errorf("form.Err: got %v, want %v", err, test.err)
+					}
+
+				case []bool:
+					form.Slice(&got, "pizza")
+					if diff := cmp.Diff(test.want, got); diff != "" {
+						t.Errorf("form.Slice: got %v, want %v", got, test.want)
+					}
+					if err = form.Err(); test.err.Error() != err.Error() {
+						t.Errorf("form.Err: got %v, want %v", err, test.err)
+					}
+				}
+				return Result{}
+			}, &dispatcher{})
+			recorder := httptest.NewRecorder()
+			m.HandleRequest(recorder, req)
+			if respStatus := recorder.Result().Status; respStatus != status200OK {
+				t.Errorf("response status: got %s, want %s", respStatus, status200OK)
+			}
+		}
+	}
+}
+
+func TestFormErrorHandling(t *testing.T) {
+	test := struct {
+		name string
+		reqs *[]*http.Request
+		errs []error
+	}{
+		name: "Erros occuring in requests",
+		reqs: func() *[]*http.Request {
+			getReq := httptest.NewRequest("GET", "/?pizzaInt=diavola&pizzaBool=true&pizzaUint=-13", nil)
+			postReq := httptest.NewRequest("POST", "/", strings.NewReader("pizzaInt=diavola&pizzaBool=true&pizzaUint=-13"))
+			postReq.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+			multipartReqBody := "--123\r\n" +
+				"Content-Disposition: form-data; name=\"pizzaInt\"\r\n" +
+				"\r\n" +
+				"diavola\r\n" +
+				"--123\r\n" +
+				"Content-Disposition: form-data; name=\"pizzaBool\"\r\n" +
+				"\r\n" +
+				"true\r\n" +
+				"--123\r\n" +
+				"Content-Disposition: form-data; name=\"pizzaUint\"\r\n" +
+				"\r\n" +
+				"-13\r\n" +
+				"--123--\r\n"
+			multipartReq := httptest.NewRequest("POST", "/", strings.NewReader(multipartReqBody))
+			multipartReq.Header.Set("Content-Type", `multipart/form-data; boundary="123"`)
+			res := []*http.Request{getReq, postReq, multipartReq}
+			return &res
+		}(),
+		errs: []error{errors.New(`strconv.ParseInt: parsing "diavola": invalid syntax`), errors.New(`strconv.ParseUint: parsing "-13": invalid syntax`)},
+	}
+
+	for _, req := range *test.reqs {
+		m := NewMachinery(func(rw ResponseWriter, ir *IncomingRequest) Result {
+			form, err := getParsedForm(ir)
+			if err != nil {
+				t.Fatalf(`getParsedForm: got "%v", want nil`, err)
+			}
+			var wantInt int64 = 0
+			gotInt := form.Int64("pizzaInt", 0)
+			if diff := cmp.Diff(wantInt, gotInt); diff != "" {
+				t.Errorf("form.Int64: got %v, want %v, diff (-want +got): \n%s", gotInt, wantInt, diff)
+			}
+			if err = form.Err(); test.errs[0].Error() != err.Error() {
```

Comment:
nit: here and elsewhere: use := where you can, to reduce the scope of variables as much as possible

Reply:
Thanks :)